src/nikonmn_int.cpp

```diff
@@ -2963,70 +3204,156 @@ std::ostream& Nikon3MakerNote::printExternalFlashData1(std::ostream& os, const V
-      break;
-    case 6:
-      os << _("Not ready");
-      break;
-    default:
-      os << "(" << temp << ")";
-      break;
+  if (v0 & 0x80) {
+    os << ", ";
+    printTag<std::size(nikonFlashControlMode), nikonFlashControlMode>(os, (value.toInt64() & 0x0F), metadata);
   }
-
   os.copyfmt(oss);
   os.flags(f);
   return os;
 }
 
-std::ostream& Nikon3MakerNote::printFlashCompensation(std::ostream& os, const Value& value, const ExifData*) {
+std::ostream& Nikon3MakerNote::printExternalFlashData1Fl7(std::ostream& os, const Value& value,
+                                                          const ExifData* metadata) {
   std::ios::fmtflags f(os.flags());
-  if (value.count() != 1 || value.typeId() != signedByte) {
+  if (value.count() != 1 || value.typeId() != unsignedByte) {
     os << "(" << value << ")";
     os.flags(f);
     return os;
   }
   std::ostringstream oss;
   oss.copyfmt(os);
-  float temp = (value.toFloat() / static_cast<float>(-6.0));
+  const auto v0 = value.toInt64();
+  os << (v0 & 0x01 ? _("External flash on") : _("External flash off"));
+
+  if (v0 & 0x01) {
+    os << ", ";
+    os << (v0 & 0x80 ? _("External flash zoom override") : _("No external flash zoom override"));
+
+    std::ostringstream ossTemp;
+    printTagBitmask<std::size(nikonFlashAdaptors), nikonFlashAdaptors>(ossTemp, value, metadata);
+    std::string tempStr = ossTemp.str();
+    if (!tempStr.empty()) {
+      os << ", " << tempStr;
+    }
+  }
+  os.copyfmt(oss);
+  os.flags(f);
+  return os;
+}
 
-  if (temp == 0)
-    os << 0;
-  else if (!std::isfinite(temp))
+std::ostream& Nikon3MakerNote::printExternalFlashData2(std::ostream& os, const Value& value, const ExifData* metadata) {
+  std::ios::fmtflags f(os.flags());
+  if (value.count() != 1 || value.typeId() != unsignedByte) {
     os << "(" << value << ")";
-  else if (std::abs(std::remainderf(temp, 1)) < 0.001)
-    os << std::round(temp);
-  else if (std::abs(std::remainderf(temp * 2, 1)) < 0.001)
-    os << std::round(temp * 2) << "/2";
-  else if (std::abs(std::remainderf(temp * 3, 1)) < 0.001)
-    os << std::round(temp * 3) << "/3";
-  else
-    os << std::setprecision(3) << temp;
+    os.flags(f);
+    return os;
+  }
+
+  printTag<std::size(nikonFlashControlMode), nikonFlashControlMode>(os, (value.toInt64() & 0x0F), metadata);
+
+  os.flags(f);
+  return os;
+}
+
+std::ostream& Nikon3MakerNote::printFlashMasterDataFl6(std::ostream& os, const Value& value, const ExifData* metadata) {
+  std::ios::fmtflags f(os.flags());
+  if (value.count() != 1 || value.typeId() != unsignedByte) {
+    os << "(" << value << ")";
+    os.flags(f);
+    return os;
+  }
+
+  // Check if using an automated or manual mode
+  auto pos = metadata->findKey(ExifKey("Exif.NikonFl6.ExternalFlashData1"));
+  if (pos == metadata->end() || pos->count() != 1 || pos->typeId() != unsignedByte) {
+    os << "(" << value << ")";
+    os.flags(f);
+    return os;
+  }
+
+  const auto mode = pos->toInt64(0) & 0x0F;
```

Comment:
Can you use `toUint32()` here? I think it's cleaner to use unsigned types for bitwise operations like `&`.

Reply:
Thanks, I have replaced all the `toInt6()` in my code as the tags use unsigned values.